### PR TITLE
fix: Correctly encode user given content, such as passwords

### LINKED ISSUE #219: Correctly encode user given content, such as passwords
https://druid.apache.org/docs/latest/configuration/#configuration-interpolation

Usage example

```
ssl.keystore.password: '${env:SSL_KEYSTORE_PASSWORD}'
ssl.truststore.password: '${env:SSL_TRUSTSTORE_PASSWORD}'
```

We can use this to replace these two usages of `sed`:
- https://github.com/stackabletech/druid-operator/blob/f5cc80d588bc110a899e333560054a2cd2364ecc/rust/crd/src/lib.rs#L567
- https://github.com/stackabletech/druid-operator/blob/f5cc80d588bc110a899e333560054a2cd2364ecc/rust/crd/src/authentication/ldap.rs#L192

### PR DESCRIPTION
**Pull request by fhennig**
# Description

fixes #219 

Integration tests were successful: https://ci.stackable.tech/view/06%20Replicated/job/druid-operator-it-replicated/8/console

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: rust/crd/src/authentication/ldap.rs [hunk 1]
```diff
@@ -183,27 +181,16 @@ impl DruidLdapSettings {
...
-        commands
-                .push(format!("echo \"Replacing LDAP placeholders with their proper values in {runtime_properties_file}\""));
-        commands.push(format!(
-            r#"sed "s|{PLACEHOLDER_INTERNAL_CLIENT_PASSWORD}|{internal_client_password}|g" -i {runtime_properties_file}"# // using another delimiter (|) here because of base64 string
-        ));
-
         if let Some((ldap_bind_user_path, ldap_bind_password_path)) =
             self.ldap.bind_credentials_mount_paths()
         {
             let ldap_bind_user = format!("$(cat {ldap_bind_user_path})");
```
Comment (sbernauer) [inline rust/crd/src/authentication/ldap.rs:187]: As discussed offline
```
            commands.push(format!(
                "export {ENV_LDAP_BIND_USER}=\"$(cat {ldap_bind_user_path})\""
            ));
            commands.push(format!(
                "export {ENV_LDAP_BIND_PASSWORD}=\"$(cat {ldap_bind_password_path})\""
            ));
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
Can you please also try to break the integration tests by setting as many weird characters in e.g. ldap bind credentials or internal passwords?

File: rust/crd/src/authentication/ldap.rs (hunk 1 above)
Comment (fhennig) [reply rust/crd/src/authentication/ldap.rs]: https://github.com/stackabletech/druid-operator/pull/571/commits/adc870f205d44becf97310da2df057452373b4bd

**Comment by fhennig** [comment]
running a [test](https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/druid-operator-it-custom/146/) :crossed_fingers:

File: rust/crd/src/authentication/ldap.rs [hunk 2]
```diff
@@ -183,27 +181,15 @@ impl DruidLdapSettings {
...
         if let Some((ldap_bind_user_path, ldap_bind_password_path)) =
             self.ldap.bind_credentials_mount_paths()
         {
-            let ldap_bind_user = format!("$(cat {ldap_bind_user_path})");
-            let ldap_bind_password = format!("$(cat {ldap_bind_password_path})");
-
             commands.push(format!(
-                    r#"sed "s/{PLACEHOLDER_LDAP_BIND_USER}/{ldap_bind_user}/g" -i {runtime_properties_file}"#
-                ));
+                "export {ENV_LDAP_BIND_USER}=\"$(cat {ldap_bind_user_path})\""
```
Comment (sbernauer) [inline rust/crd/src/authentication/ldap.rs:188]: TIL: Actually, we should use something like `druid.segmentCache.locations=${file:UTF-8:/config/segment-cache-def.json}` if possible (see https://druid.apache.org/docs/latest/configuration/#configuration-interpolation)

Comment (fhennig) [reply rust/crd/src/authentication/ldap.rs:188]: oh cool! I should've read the docs better haha

Comment (fhennig) [reply rust/crd/src/authentication/ldap.rs]: https://github.com/stackabletech/druid-operator/pull/571/commits/9144aa2f9155c3a3833e4cd8544a0b249b27a2d5 
https://github.com/stackabletech/druid-operator/pull/571/commits/6f8f24177e2a83228ff9a729bd6072b87c950c31

**Review by sbernauer — DISMISSED** [review]
Thanks! LGTM when tests pass

File: rust/operator-binary/src/druid_controller.rs [hunk 3]
```diff
@@ -711,6 +712,21 @@ fn build_rolegroup_config_map(
...
+                        conf.insert(
+                            S3_ACCESS_KEY.to_string(),
+                            Some(format!(
+                                "${{file:UTF-8:{S3_SECRET_DIR_NAME}/{SECRET_KEY_S3_ACCESS_KEY}}}"
+                            )),
+                        );
+                        conf.insert(
+                            S3_SECRET_KEY.to_string(),
+                            Some(format!(
+                                "${{file:UTF-8:{S3_SECRET_DIR_NAME}/{SECRET_KEY_S3_SECRET_KEY}]]"
```
Comment (labrenbe) [inline rust/operator-binary/src/druid_controller.rs:725]: This seems to be a type and should probably read `"${{file:UTF-8:{S3_SECRET_DIR_NAME}/{SECRET_KEY_S3_SECRET_KEY}}}"`

**Review by labrenbe — CHANGES_REQUESTED** [review]

File: tests/templates/kuttl/ldap-authentication/03_openldap.yaml.j2 [hunk 4]
```diff
@@ -44,7 +44,7 @@ spec:
             - name: LDAP_ADMIN_USERNAME
               value: admin
             - name: LDAP_ADMIN_PASSWORD
-              value: admin
+              value: "admin-pw withSpace$%\" &&} §"  # use spaces and special characters to make sure it doesn't break anything
```
Comment (labrenbe) [inline tests/templates/kuttl/ldap-authentication/03_openldap.yaml.j2:47]: The password in `create_ldap_user.sh` should also be changed

**Review by labrenbe — CHANGES_REQUESTED** [review]

**Comment by sbernauer** [comment]
Just a meta comment: product-utils is coming to an release soon-ish. I think it would make sense to use that tool everywhere for consistency reasons [and not rely on Druid features that might do thing *slightly* differently].

**Comment by fhennig** [comment]
You might be right there. I can also see an argument for using less parts, but maybe the consistency is better to have here. Happy to have other opinions, maybe @soenkeliebau

File: rust/operator-binary/src/druid_controller.rs (hunk 3 above)
Comment (sbernauer) [reply rust/operator-binary/src/druid_controller.rs]: Thx, fixed in 550bd69

File: tests/templates/kuttl/ldap-authentication/03_openldap.yaml.j2 (hunk 4 above)
Comment (sbernauer) [reply tests/templates/kuttl/ldap-authentication/03_openldap.yaml.j2:47]: Actually, we should not bind from Druid with the admin password. I will update the PR to leave the admin password unchanged and instead bind with the integrationtest user (similar to https://github.com/stackabletech/nifi-operator/pull/627). Let's see how it goes

**Review by labrenbe — DISMISSED** [review]

**Review by labrenbe — APPROVED** [review]
lgtm

**Pull request merged**